### DISCUSSION TITLE
`size` with `shift` on `crossAxis` doesn't size up when shifting

### DISCUSSION THREAD

**Discussion by madoci** [floating-ui/floating-ui#3005] [Q&A] [answered]
I believe this is a bug, but I'd like your opinion before creating an issue for it.

**Describe the bug**

When using `shift` with `crossAxis` set to true, along with `size`, the `availableHeight` (or `availableWidth` if placement is left or right) doesn't increase to the actual available height when the floating element shifts over the reference.

**To Reproduce**

1. Go to https://codesandbox.io/s/modest-dream-dy5ns6?file=/src/index.js
2. Resize the window down until the tooltip hits the top and starts shrinking.
3. Resize the window up.
4. The tooltip moves down, but stays the same size. Its starts growing again when the top hits the bottom of the reference.

**Expected behavior**

When resizing up, the tooltip should regrow to fill the new available space.

**Screenshots**
![Capture d’écran du 2024-08-14 15-36-35](https://github.com/user-attachments/assets/10543973-2a7a-4103-956c-2437fc52e07c)

**Context:**

- OS: Debian
- Browser: Chrome 127.0.6533.99
- Version: @floating-ui/dom 1.6.10

**Additional context**

I think the issue lies here : 

https://github.com/floating-ui/floating-ui/blob/a8b7a88a73f780c59b89bb8c6e3db16ec589a43d/packages/core/src/middleware/size.ts#L85-L95

The `availableHeight` or `availableWidth` is set to `maximumClippingHeight` when shifting, but only on the main shift axis.
In fact, this code seems overcomplicated to me. The `overflowAvailable<size>` is set before to the min between some value and `maximumClipping<size>`, so `min(overflowAvailable<size>, maximumClipping<size>)` is equivalent to `overflowAvailable<size>`. And since `available<size>` is set before to `overflowAvailable<size>`, the ternary condition essentially means `If alignment or noShift, do nothing. Otherwise, set to maximumClipping<size>`. So the condition could be put in the `if... else...`. It doesn't change much, but it will allow me to simplify my proposal.

The `available<size>` should be set to `maximumClipping<size>` if there is a shift along its axis, even if it's on the placement axis. Something like this would work : 
```js
if (state.middlewareData.shift?.x && !(alignment && isYAxis)) {
  availableHeight = maximumClippingHeight;
}
if (state.middlewareData.shift?.y && !(alignment && !isYAxis)) {
  availableHeight = maximumClippingHeight;
}
```

This works in my case, but I don't know the side effects it can have.

However, during testing on the codesandbox, I noticed that the behaviour also depends on the padding given to `shift` and `size`. The behaviour explained here shows when we use the same padding for both middleware. In the codesanbox, you can clik on the "Show other tooltips" button to display a tooltip where `size` has less padding than `shift`, and one where `size` has more padding than `shift`. With less padding, the tooltip resize almost correctly (it has some weird deferment when approaching its max height). With more padding, the tooltip shifts only when it cannot shrink anymore.

This appears to come from what `detectOverflow` returns. With the code above, all of these cases seems to be fixed, but I don't know if this is a bug of the `detectOverflow` function or if it works as intended.

**Comment by atomiks** [comment, marked as answer]
> This works in my case, but I don't know the side effects it can have.

It's fine to submit a PR as there are lots of functional tests for this behavior (`shift` + `size` enabled/disabled combinations), and the screenshots can inform you what's happened with your change. The current code probably handles cases where one is present but not the other, and vice-versa.

**Comment by madoci** [comment]
Created issue #3011
